garage/tf/models/base.py

```diff
@@ -0,0 +1,51 @@
+"""This file contains the tautology for garage models."""
+import tensorflow as tf
+from tensorflow.keras.models import Model as KerasModel
+
+from garage.misc.overrides import overrides
+from garage.tf.core import Parameterized
+
+
+class Model(KerasModel, Parameterized):
```

Comment:
I have a few high level questions about this class, which are hinted at in the discussion below, but don't look like they are being discussed directly:
 - Do we still need parameterized if we're using KerasModel, or does the Keras API provide sufficient functionality?
 - Do we really want to put Parameterized (a subclass of Serializable) this far up the inheritance heirarchy? Note that the conclusion from the discussion [here](https://github.com/rlworkgroup/proposals/pull/3) was specifically not to do that.
 - If we're using the Keras API explicitly, how much do we want to follow their naming conventions? Should we ensure that we use all of their names, or do we prefer to use our own?

Reply:
I agree with KR's questions. Part of the hope here is to removed Parameterized.

I think we should follow the "when in rome" principle and adopt their naming conventions.

@krzentner can you give an example of their naming convention on which we might differ?

Reply:
- Keras Model API now doesn't provide serializable mothods for subclassing models. 

- Keras layers support `layer.get_config` and `layer.get_weights` for serialization. In order to use layers methods, we need access to layers in the network. i.e.
``` python
class Model(KerasModel):
    def __getstate__(self,):
        weights = [layer.get_weights() for layer in model.layers]
```
In garage, Model calls mlp or other networks in `garage.tf.core` to build its network. According to Keras documentation, I don't think we can keep track of layers here unless we make network to a class.

Reply:
I also have some questions here:
- Is the purpose of Model only to use serialization in Keras? 
- After reading the Keras Model doc, it looks like they are useful to compose layers for complex networks. Is this the same idea as Model in garage? i.e. According to the docstring below, we can assemble multiple models in garage, but Keras Model doesn't support the same idea.

Reply:
Another concern on removing `Parameterized` here:
Some optimizers in `garage.tf.optimizers` uses `policy.get_param_values`. 
Optimizers in TF:
- ConjugateGradientOptimizer.
- LBFGSOptimizer. There is one [BFGSOptimizer](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/optimizer/bfgs.py) in tf-prob.
- FirstOrderOptimizer, which is basically Adam.
- PenaltyLbfgsOptimizer. 

Can we find alternative optimizers in TF to replace them?

Reply:
@CatherineSue re: Graph Networks vs Subclass Networks.

This is certainly disappointing, and I think there are a couple ways to deal with it.

Firstly, I think their explanation is a bit unclear. I think that they are saying is: if you *simply* subclass `Model` *without* using the Keras composition APIs to build your model, then you won't get serialization *automatically*.

You are of course free to implement those methods yourself--it's still Python after all.

This leaves us with 2 choices:
1. Implement all primitives using Keras-style graph composition (Sequential and Functional APIs) -- this is basically a "Layers" library
1. Implement our own subclass(es) of Model which provide `clone()`, `get_config()`/`from_config()`, `to_json()`/`to_yaml()`, and `save()` -- this is basically what we have now with `Serializable` and `Parameterized`, but they are separate classes rather than being part of one Model API

If we implement the Keras API (using either 1 or 2), then future users will have both options (they can mix/match models built using the Keras builder APIs with our own).

Now for what to do. This is not 100% clear.

* I think that if the Keras Sequential/Functional APIs suit our needs, and implementing custom Layers is possible and easy, and that keras.Model will properly save Models with custom layers, we should implement our own Layers and then do 1.
* If the above does not hold, we should probably do (2).

Justification:
* We satisfy the original goal or removing Parameterized and Serializable
* In the worst case, it seems to me you can always make one giant layer for your whole primitive and then have a Model with one layer (not ideal of course, but possible)
* We are already implementing "Layers" with our network functions. This would just make the functions we already have into classes.
* Since TF has committed to the Keras API, it's likely that support for custom models will get better in the future
* We might be able to reduce our testing surface by using Layers already provided by Keras
* We get to keep the original intent of the Layers library without having to maintain any of it.

Reply:
I tested it and `keras.Model` does not properly save Models with primitives using Keras-style graph composition. 
The only condition that `keras.Model` save the model is that the model is created by Sequential/Functional API, i.e. either `keras.models.Sequential[...]` or `keras.model.Model(inputs=..., outputs=...)`